qiskit/transpiler/passes/layout/sabre_layout.py

```diff
@@ -227,59 +230,84 @@ def run(self, dag):
+            {
+                component.dag.qubits[logic]: component.coupling_map.graph[phys]
+                for component in components
+                for logic, phys in component.initial_layout.layout_mapping()
+                # A physical component of the coupling map might be wider than the DAG that we're
+                # laying out onto it.  We shouldn't include these implicit ancillas right now as the
+                # ancilla-allocation pass will run on the whole map in one go.
+                if logic < len(component.dag.qubits)
+            }
         )
-        initial_layout_dict = {}
-        final_layout_dict = {}
-        for (
-            layout_dict,
-            final_dict,
-            component_map,
-            _sabre_result,
-            _circuit_to_dag_dict,
-            local_dag,
-        ) in layout_components:
-            initial_layout_dict.update({k: component_map[v] for k, v in layout_dict.items()})
-            final_layout_dict.update({component_map[k]: component_map[v] for k, v in final_dict})
-        self.property_set["layout"] = Layout(initial_layout_dict)
+
         # If skip_routing is set then return the layout in the property set
         # and throwaway the extra work we did to compute the swap map.
         # We also skip routing here if there is more than one connected
         # component we ran layout on. We can only reliably route the full dag
         # in this case if there is any dependency between the components
         # (typically shared classical data or barriers).
-        if self.skip_routing or len(layout_components) > 1:
+        if self.skip_routing or len(components) > 1:
             return dag
-        # After this point the pass is no longer an analysis pass and the
-        # output circuit returned is transformed with the layout applied
-        # and swaps inserted
-        dag = self._apply_layout_no_pass_manager(dag)
-        mapped_dag = dag.copy_empty_like()
+
+        # At this point, we become a transformation pass, and apply the layout and the routing to
+        # the DAG directly.  This includes filling in the `property_set` data of the embed passes.
+
+        dag = self._ancilla_allocation_no_pass_manager(dag)
+        # The ancilla-allocation pass has expanded this since we set it above.
+        full_initial_layout = self.property_set["layout"]
+
+        # Set up a physical DAG to apply the Sabre result onto.  We do not need to run the
+        # `ApplyLayout` transpiler pass (which usually does this step), because we're about to apply
+        # the layout and routing together as part of resolving the Sabre result.
+        physical_qubits = QuantumRegister(self.coupling_map.size(), "q")
+        mapped_dag = DAGCircuit()
+        mapped_dag.add_qreg(physical_qubits)
+        mapped_dag.add_clbits(dag.clbits)
+        for creg in dag.cregs.values():
+            mapped_dag.add_creg(creg)
+        mapped_dag._global_phase = dag._global_phase
```

Comment:
I assume you're using the inner private attrs here to avoid the input type checking on the public setter?

Reply:
I don't think I touched this line other than the variable renaming, but that's probably the root reason - the setter does normalisation to `[0, 2pi)` as well.